pkg/controllers/flannelmigration/network_migrator.go

```diff
@@ -0,0 +1,191 @@
+	pod := k8spod("remove-flannel")
+	podLog, err := pod.RunPodOnNodeTillComplete(m.k8sClientset, namespaceKubeSystem, m.calicoImage, node.Name, cmd, m.config.CNIConfigDir, true, true)
+	if podLog != "" {
+		log.Infof("remove-flannel pod logs: %s.", podLog)
+	}
+
+	// Wait twice as long as default sync period so kubelet has picked up dummy CNI config.
+	// TODO we probably need something better here.
+	time.Sleep(2 * defaultSyncConfigPeriod)
+	return err
+}
+
+// Drain node, remove Flannel and setup Calico network for a node.
+func (m *networkMigrator) setupCalicoNetworkForNode(node *v1.Node) error {
+	log.Infof("Setting node label to disable Flannel daemonset pod on %s.", node.Name)
+	// Set node label so no Flannel pod can be scheduled on this node.
+	// Flannel pod currently running on the node starts to be evicted as the side effect.
+	n := k8snode(node.Name)
+	err := n.addNodeLabels(m.k8sClientset, nodeNetworkUnknown)
+	if err != nil {
+		log.WithError(err).Errorf("Error adding node label to disable Flannel network for node %s.", node.Name)
+		return err
+	}
+
+	// Cordon and Drain node. Make sure no pod (except daemonset pod or pod with nodeName selector) can run on this node.
+	// TODO
+
+	// Wait for flannel pod to disappear before we proceed, otherwise it may reinstall Flannel network.
+	//n.waitPodsDisappearForNode(m.k8sClientset, 1*time.Second, 2*time.Minute, func(pod *v1.Pod) bool {
+	//	return true
+	//})
+
+	log.Infof("Removing flannel tunnel device/routes on %s.", node.Name)
+	// Remove Flannel network from node.
+	// Note Flannel vxlan tunnel device (flannel.1) is created by Flannel daemonset pod (not Flannel CNI)
+	// Therefor if Flannel daemonset pod can not run on this node, the tunnel device will not be recreated
+	// after we delete it.
+	err = m.removeFlannelNetworkAndInstallDummyCalicoCNI(node)
+	if err != nil {
+		log.WithError(err).Errorf("failed to remove flannel network on node %s", node.Name)
+		return err
+	}
+
+	log.Infof("Deleting non-host-networked pods on %s.", node.Name)
+	// Delete all pods on the node which is not host networked.
+	err = n.deletePodsForNode(m.k8sClientset, func(pod *v1.Pod) bool {
+		return pod.Spec.HostNetwork == false
+	})
+	if err != nil {
+		log.WithError(err).Errorf("failed to delete non-host-networked pods on node %s", node.Name)
+		return err
+	}
+
+	// Set node label so that Calico daemonset pod start to run on this node.
+	// This will install Calico CNI configuration file.
+	// It will take the preference over Flannel CNI config or Canal CNI config.
+	log.Infof("Setting node lable to enable Calico daemonset pod on %s.", node.Name)
+	err = n.addNodeLabels(m.k8sClientset, nodeNetworkCalico)
+	if err != nil {
+		log.WithError(err).Errorf("Error adding node label to enable Calico network for node %s.", node.Name)
+		return err
+	}
+
+	// Calico daemonset pod should start running now.
+	// TODO do we need to wait for calico node pod ready?
```

Comment:
I think we should - if we accidentally roll out a bad configuration we don't want to continue.

Reply:
Agree. Code added to wait for calico-node pod running.